Spatial_Engine/Query/DominantVector.cs

```diff
@@ -0,0 +1,142 @@
+        }
+
+        /******************************************/
+        /****            IElement2D            ****/
+        /******************************************/
+        
+        [Description("Gets the the dominant vector (orientation) of an Element2D based on its lines lengths.")]
+        [Input("element1D", "Element2D to evaluate.")]
+        [Input("orthogonalPriority", "Optional, if true gives priority to curves that are on the orthogonal axis (X, Y or Z vectors.")]
+        [Input("orthogonalLengthTolerance", "Optional, tests the orthogonal vector length's in relation to the actual non-orthogonal dominant vector. For example if the dominant vector is 10 in length but the orthogonal is only 5 in length, then this number should be 0.5 for it to pass the test.")]
+        [Input("angleTolerance", "Optional, angle in radians that vectors will be considered similar.")]
+        [Output("dominantVector", "The dominant vector of an Element2D.")]
+        public static BH.oM.Geometry.Vector DominantVector(IElement2D element2D, bool orthogonalPriority = true, double orthogonalLengthTolerance = 0.5, double angleTolerance = BH.oM.Geometry.Tolerance.Angle)
+        {
+            IElement1D outline = BH.Engine.Geometry.Create.PolyCurve(element2D.IOutlineElements1D().Select(x =>x.IGeometry()));
+            return DominantVector(outline, orthogonalPriority, orthogonalLengthTolerance, angleTolerance);
+        }
+        
+        /******************************************/
+        /****              Private             ****/
+        /******************************************/
+
+        [Description("Groups vectors by direction whilst allowing for an angle discrepancy tolerance.")]
+        [Input("vectors", "Vectors to evaluate.")]
+        [Input("angleTolerance", "The angle in radians to compare vectors with each other for tolerance when grouping.")]
+        [Output("GroupSimilarVectorsWithTolerance", "The grouped vectors.")]
+        private static List<List<Vector>> GroupSimilarVectorsWithTolerance(List<Vector> vectors, double angleTolerance)
+        {
+            List<List<Vector>> result = new List<List<Vector>>();
+            List<Vector> orderByLength = vectors.OrderByDescending(x => x.Length()).ToList();
+
+            for (int i = 0; i < orderByLength.Count; i++)
+            {
+                List<Vector> sublist = new List<Vector>();
+                sublist.Add(orderByLength[i]);
+                
+                for (int j = 1; j < orderByLength.Count; j++)
+                {
+                    if (orderByLength[i].IsParallel(orderByLength[j],angleTolerance) != 0)
+                    {
+                        sublist.Add(orderByLength[j]);
+                        orderByLength.RemoveAt(j);
+                        j = j - 1;
+                    }
+                }
+                orderByLength.RemoveAt(i);
+                i = i - 1;
+                result.Add(sublist);
+            }
```

Comment:
I had another brief look at this loop, it definitely falls into `while` crowd rather than `for`, plus removal of elements from collections infers decrementation. I gave it a shot below. Just in case: nothing wrong with your code, I just felt a Codewars spirit for a second 😃 
```suggestion
            while (orderByLength.Count != 0)
            {
                List<Vector> sublist = new List<Vector>();
                sublist.Add(orderByLength[0]);
                
                for (int i = orderByLength.Count - 1; i > 0; i--)
                {
                    if (orderByLength[0].IsParallel(orderByLength[i],angleTolerance) != 0)
                    {
                        sublist.Add(orderByLength[i]);
                        orderByLength.RemoveAt(i);
                    }
                }
                orderByLength.RemoveAt(0);
                result.Add(sublist);
            }
```